gcc/common/config/riscv/riscv-common.c

```diff
@@ -441,29 +489,215 @@ riscv_subset_list::parse_sv_or_non_std_ext (const char *p,
       while (*++q != '\0' && *q != '_' && !ISDIGIT (*q))
 	;
 
-      end_of_version
-	= parsing_subset_version (q, &major_version, &minor_version,
-				  /* default_major_version= */ 2,
-				  /* default_minor_version= */ 0,
-				  /* std_ext_p= */ FALSE);
+      end_of_version = parsing_subset_version (q, &major_version, &minor_version,
+					       /* default_major_version= */ 2,
+					       /* default_minor_version= */ 0,
+					       /* std_ext_p= */ false);
 
       *q = '\0';
 
+      /* Check that the name is valid.
+	 For 'x', anything goes but it cannot simply be 'x'.
+	 For 'z', it must be known from a list and also cannot simply be 'z'.
+	 For 's', it must be known from a list and also *can* simply be 's'.
+	 For 'sx', see 'x'.  */
+
+      /* Check that the extension name is well-formed.  */
+      if (!config->ext_valid_p (subset))
+	{
+	  error_at (m_loc, "%<-march=%s%>: Invalid or unknown %s ISA extension: %qs",
+		    m_arch, config->prefix_upper, subset);
+	  free (subset);
+	  return NULL;
+	}
+
+      /* Check that the last item is not the same as this.  */
+      last_name = m_tail->name.c_str();
+
+      if (!strcasecmp (last_name, subset))
+	{
+	  error_at (m_loc, "%<-march=%s%>: Duplicate %s ISA extension: %qs",
+		    m_arch, config->prefix_upper, subset);
+	  free (subset);
+	  return NULL;
+	}
+
+      /* Check that we are in alphabetical order within the subset.  */
+      if (!strncasecmp (last_name, config->prefix_lower, 1)
+	  && strcasecmp (last_name, subset) > 0)
+	{
+	  error_at (m_loc, "%<-march=%s%>: %s ISA extension not in alphabetical order: "
+		    "%qs must come before %qs.",
+		    m_arch, config->prefix_upper, subset, last_name);
+	  free (subset);
+	  return NULL;
+	}
+
       add (subset, major_version, minor_version);
       free (subset);
       p += end_of_version - subset;
 
       if (*p != '\0' && *p != '_')
 	{
-	  error_at (m_loc, "%<-march=%s%>: %s must separate with _",
-		    m_arch, ext_type_str);
+	  error_at (m_loc, "%<-march=%s%>: %qs must seperate with _",
+		    m_arch, config->prefix_lower);
 	  return NULL;
 	}
     }
 
   return p;
 }
 
+#define RISCV_STD_Z_EXT_COUNT 9
+
+const char * const riscv_std_z_ext_strtab[RISCV_STD_Z_EXT_COUNT] =
+  {
+   "zbb", "zbc", "zbe", "zbf", "zbm", "zbp", "zbr", "zbs", "zbt"
+  };
+
+#define RISCV_STD_S_EXT_COUNT 0
+
+const char * const riscv_std_s_ext_strtab[RISCV_STD_S_EXT_COUNT] =
+  {
+
+  };
+
+/* Find the index corresponding to the z-extension
+   name `ext'. If not found, return -1.  */
+
+int
+riscv_std_z_ext_index (const char *ext)
+{
+  int i;
+
+  for (i = 0; i < RISCV_STD_Z_EXT_COUNT; ++i)
+    {
+      if (!strcasecmp (ext, riscv_std_z_ext_strtab[i]))
+	return i;
+    }
+
+  return -1;
+}
+
+/* Predicator function for x-prefixed extensions.
+   Anything goes, except the literal 'x'.  */
+
+static int
+riscv_ext_x_valid_p (const char *arg)
```

Comment:
`riscv_ext_*_valid_p` could be removed or simplified once we have ext_info_t table.